### PR TITLE
Adding Hisat2

### PR DESCRIPTION
**Pull request by arahuja**
This has been tested insofar in that it correctly produces the index and proper hisat2 command (alongside with hisat1) but the hisat2 alignment has not completed. I figured it was worth reviewing in the meantime

This follows the model we discussed previously:
- Creates a versioned Hisat in `run_environment.ml`
- Add a version argument to `hisat` in `tool_providers.ml` as well as putting the version in the install path
- Adds config to hisat to switch between versions

Let me know any critiques or if anything can be improved.

cc @smondet @ihodes

---
This change is [Reviewable](https://reviewable.io/reviews/hammerlab/biokepi/219)

### CODE REVIEW

File: src/lib/run_environment.ml [hunk 1]
```diff
@@ -54,7 +55,7 @@ module Tool = struct
     let star = custom "star" ~version:"2.4.1d"
     let stringtie = custom "stringtie" ~version:"1.2.2"
     let cufflinks = custom "cufflinks" ~version:"2.2.1"
-    let hisat = custom "hisat" ~version:"0.1.6-beta"
+    let hisat = `Hisat `V_2_0_2_beta
```
Comment (ihodes) [inline src/lib/run_environment.ml]: I think this is good; latest tools should be the default.

File: src/lib/tool_providers.ml [hunk 2]
```diff
@@ -494,7 +506,7 @@ let default_toolkit
     star_tool ~host ~meta_playground;
     stringtie_tool ~host ~meta_playground;
     cufflinks_tools ~host ~meta_playground;
-    hisat_tool ~host ~meta_playground;
+    hisat_tool ~host ~meta_playground ~version:`V_0_1_6_beta;
```
Comment (ihodes) [inline src/lib/tool_providers.ml]: This is my confusion around tools again; why V2 in the run_environment but V1 in the default toolkit?

Comment (arahuja) [inline src/lib/tool_providers.ml]: Yea, actually was just thinking that - the two are not really linked. Here I create `V_0_1_6_beta` tool, and there I say if you ask for the default it's the V2 one (which may not be in the kit unless you add it)

Comment (ihodes) [inline src/lib/tool_providers.ml]: Maybe have both tools in the toolkit, but the default tool should remain V2?

Comment (arahuja) [inline src/lib/tool_providers.ml]: I guess the question is should there be a 1-1 mapping between the default tools and the default toolkit - if so, how is it enforced

Comment (arahuja) [inline src/lib/tool_providers.ml]: > Maybe have both versions in the toolkit, but the default tool should remain V2?

I thought have 2 versions of a single tool in the default kit seemed weird, I added V2 to the kit when I used it.

Comment (ihodes) [inline src/lib/tool_providers.ml]: I think it's fine to have other versions in the default toolkit. What seems more confusing is saying that the default tool does not exist in the default toolkit.

Comment (ihodes) [inline src/lib/tool_providers.ml]: All the nodes in the default toolkit are doing is telling the machine how to get the tool, right?

File: src/lib/tool_providers.ml [hunk 3]
```diff
@@ -111,12 +111,24 @@ let star_tool ~host ~meta_playground =
...
   let archive = Filename.basename url in
-  let hisat_binary = "hisat" in
+  let hisat_binary = 
+    match version with
+    | `V_0_1_6_beta -> "hisat"
+    | `V_2_0_2_beta -> "hisat2"
+  in
   let ensure =
     workflow_node (single_file ~host (install_path // hisat_binary))
       ~name:(sprintf "Install HISAT")
```
Comment (ihodes) [inline src/lib/tool_providers.ml]: Maybe we should add the version to the name?

File: src/lib/pipeline.ml [hunk 4]
```diff
@@ -158,7 +158,7 @@ module Construct = struct
 
   let star fastq = Star fastq
 
-  let hisat fastq = Hisat fastq
+  let hisat ?(configuration = Hisat.Configuration.default_v1) fastq = Hisat(configuration, fastq)
```
Comment (smondet) [inline src/lib/pipeline.ml]: Shouldn't `v2` be the default?

File: src/lib/tool_providers.ml (hunk 2 above)
Comment (smondet) [inline src/lib/tool_providers.ml]: Yes, having both version in the default tool-kit is what we want. They will be installed only when used anyway.

File: src/lib/pipeline.ml (hunk 4 above)
Comment (arahuja) [inline src/lib/pipeline.ml]: If you'd like? My assumption was the default should be what was there already so nothing changes for any pipelines using v1

Comment (smondet) [inline src/lib/pipeline.ml]: Actually maybe :)

I don't know how the two versions are different, if you think users expect to very different results then yes `v1` is better.

**Comment by arahuja** [comment]
@smondet @ihodes not sure if you are still reviewing, but from the feedback so far, my takeaways are:

1. Add both to the Toolkit.default
2. Make the default tool in run environment and default config in pipeline match and that should be ... `V2`?

**Comment by ihodes** [comment]
1. for sure (no downside)
2. i don't have a preference; seems good to keep things recent, but also have to watch out for that e.g. I need to switch the PGV pipeline code if we make this change.

**Comment by arahuja** [comment]
> i don't have a preference; seems good to keep things recent, but also have to watch out for that e.g. I need to switch the PGV pipeline code if we make this change.
@ihodes For this case, it doesn't matter really as we likely know all end-users of this, but in general, I think major upgrades (especially when the previous version isn't broken) should be opt-in rather than opt-out.  In this case, if you using `hisat-v1` I think it's worthwhile to stay with until we have some time to sanity check and compare v2

**Comment by smondet** [comment]
@arahuja yes to both !

**Pull request merged**